app/utils/GatekeeperAuthWrapper.scala

```diff
@@ -16,52 +16,47 @@
 
 package utils
 
+import config.AppConfig
 import connectors.AuthConnector
-import controllers.BaseController
-import model.GatekeeperRole
+import model.{GatekeeperRole, LoggedInUser, LoggedInRequest}
 import model.GatekeeperRole.GatekeeperRole
-import play.api.Play.current
-import play.api.i18n.Messages.Implicits._
-import play.api.mvc.{Action, AnyContent, Request, Result, _}
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, Request, Result}
 import uk.gov.hmrc.auth.core._
 import uk.gov.hmrc.auth.core.authorise.Predicate
 import uk.gov.hmrc.auth.core.retrieve.{~, _}
-import uk.gov.hmrc.http.HeaderCarrier
-import uk.gov.hmrc.play.HeaderCarrierConverter
+import uk.gov.hmrc.play.bootstrap.controller.FrontendBaseController
+import views.html.ForbiddenView
 
-import scala.concurrent.Future
-
-case class LoggedInUser(userFullName: Option[String])
-
-trait GatekeeperAuthWrapper {
-  self: BaseController =>
+import scala.concurrent.{ExecutionContext, Future}
 
+trait GatekeeperAuthWrapper extends I18nSupport{
+  self: FrontendBaseController =>
   def authConnector: AuthConnector
 
   implicit def loggedIn(implicit request: LoggedInRequest[_]): LoggedInUser = LoggedInUser(request.name)
 
-  def requiresAtLeast(minimumRoleRequired: GatekeeperRole)(body: LoggedInRequest[_] => HeaderCarrier => Future[Result]): Action[AnyContent] = Action.async {
+  def requiresAtLeast(minimumRoleRequired: GatekeeperRole, forbiddenView: ForbiddenView)(body: LoggedInRequest[_] => Future[Result])
```

Comment:
Should (Can?) the forbiddenView be moved to a field that the implementor has to provide? Or are there cases where different actions use a different view? Instead of always passing it in.

Reply:
I think is one of the problems we faced in this upgrade - the actions going off to views when things go wrong.
Because the views are DI'd and because these actions are functions on traits, we end up in a difficult place.

Reply:
We have altered others to have an abstract val that needs to be implemented by the class mixing in the trait so we could do the same here - perhaps. See ErrorHelper for example.